homeassistant/components/modbus/climate.py

```diff
@@ -156,7 +157,8 @@ async def async_set_temperature(self, **kwargs):
             (kwargs.get(ATTR_TEMPERATURE) - self._offset) / self._scale
         )
         byte_string = struct.pack(self._structure, target_temperature)
-        register_value = struct.unpack(">h", byte_string[0:2])[0]
+        struct_string = f">{DEFAULT_STRUCT_FORMAT[self._data_type]}"
+        register_value = struct.unpack(struct_string, byte_string)[0]
```

Comment:
It's my understanding that pymodbus cannot handle calls to write_register with floats. It only accepts values that will fit within 2 bytes (size of a single modbus register). If we unpack with the struct format specified by `self._data_type`, then floats will still not work.

Reply:
If there was a `DEFAULT_STRUCT_REGISTER_COUNT` const dictionary something like this, it would be doable.
```
DEFAULT_STRUCT_REGISTER_COUNT = {
    DATA_TYPE_INT16: 1,
    DATA_TYPE_INT32: 2,
    DATA_TYPE_INT64: 4,
    ...
} 
```

Reply:
the user must use count: if needed, fir DATA_TYPE_STRING we cannot know the size, so the general idea is that the user knows what he/she wants.

Reply:
pymodbus handles write_register with floats very well using the array call.